## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [65]:
import pandas as pd
import numpy as np

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [66]:
# View the column names in the data
#list(df.columns)
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [67]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [68]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.count()

first          54639
last           54639
job            54639
phone          54639
email          54639
client_id      54639
order_id       54639
order_date     54639
order_week     54639
order_year     54639
item_id        54639
category       54639
subcategory    54639
unit_price     54639
unit_cost      54639
unit_weight    54639
qty            54639
line_number    54639
dtype: int64

In [69]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df =  df.sort_values('client_id', ascending=False)
df.head(10)

first    last            job               phone  \
11110  Samantha  Murray  Herpetologist  534.390.2260x56734   
151    Samantha  Murray  Herpetologist  534.390.2260x56734   
5052   Samantha  Murray  Herpetologist  534.390.2260x56734   
31669  Samantha  Murray  Herpetologist  534.390.2260x56734   
42337  Samantha  Murray  Herpetologist  534.390.2260x56734   
4053   Samantha  Murray  Herpetologist  534.390.2260x56734   
43411  Samantha  Murray  Herpetologist  534.390.2260x56734   
29231  Samantha  Murray  Herpetologist  534.390.2260x56734   
21853  Samantha  Murray  Herpetologist  534.390.2260x56734   
41584  Samantha  Murray  Herpetologist  534.390.2260x56734   

                                         email  client_id  order_id  \
11110  murray.samantha.4079@gonzalez-brown.com      99984   9673005   
151    murray.samantha.4079@gonzalez-brown.com      99984   6821353   
5052   murray.samantha.4079@gonzalez-brown.com      99984   9673005   
31669  murray.samantha.4079@gonzalez-brown.com      99984   8437912   
42337  murray.samantha.4079@gonzalez-brown.com      99984   7810921   
4053   murray.samantha.4079@gonzalez-brown.com      99984   6821353   
43411  murray.samantha.4079@gonzalez-brown.com      99984   6821353   
29231  murray.samantha.4079@gonzalez-brown.com      99984   6821353   
21853  murray.samantha.4079@gonzalez-brown.com      99984   5491843   
41584  murray.samantha.4079@gonzalez-brown.com      99984   5491843   

       order_date  order_week  order_year         item_id     category  \
11110  2023-02-27           9        2023  EUX51497-34-6B  consumables   
151    2023-02-17           7        2023  AEA17352-13-5M    furniture   
5052   2023-02-27           9        2023  ZAB46262-52-6E    furniture   
31669  2023-01-25           4        2023  ADB89726-97-4E  consumables   
42337  2023-01-16           3        2023  ABU04430-78-4B  consumables   
4053   2023-02-17           7        2023  ZAA35509-03-0U  consumables   
43411  2023-02-17           7        2023  MDZ67579-00-1X  consumables   
29231  2023-02-17           7        2023  XZZ88006-14-6D  consumables   
21853  2023-04-28          17        2023  XUC78582-36-0E  consumables   
41584  2023-04-28          17        2023  EBA49136-27-6B     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
11110   bathroom supplies      175.21     162.27        13.83   100   
151                tables      125.47      86.77         0.54   139   
5052               stands       30.21      19.49         2.62    61   
31669    kitchen supplies        8.67       6.33         2.22    53   
42337      paper products      101.40      74.52         8.37  6145   
4053     kitchen supplies        1.71       1.16         1.34   198   
43411    kitchen supplies       45.68      28.03         3.72  2067   
29231   bathroom supplies        3.23       2.10        10.01     6   
21853      paper products       82.57      59.42         0.00    20   
41584  project management       36.49      21.29        13.78    35   

       line_number  
11110            2  
151              8  
5052             0  
31669            0  
42337            1  
4053             3  
43411            4  
29231            5  
21853            5  
41584            2

In [70]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().nlargest(3)
top_categories

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [71]:
# For the category with the most entries, which subcategory had the most entries?

most_common_category = df['category'].value_counts().idxmax()
most_common_category

consum_filter_df = df[df['category'] == most_common_category]
top_subcategory = consum_filter_df['subcategory'].value_counts().nlargest(1)

top_subcategory


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [72]:
# Which five clients had the most entries in the data?
clients_5 = df['client_id'].value_counts().nlargest(5)
clients_5

client_id
33615    220
66037    211
46820    209
24741    207
38378    207
Name: count, dtype: int64

In [73]:
# Store the client ids of those top 5 clients in a list.
c_5_list= list(clients_5.keys())
c_5_list

[33615, 66037, 46820, 24741, 38378]

In [74]:
# How many total units (the qty column) did the client with the most entries order order?
#client with most entries identification
most_entries = df['client_id'].value_counts().idxmax()
most_entries

33615

In [75]:
# How many total units (the qty column) did the client with the most entries order order?
#identifying the total amount of units
units_ordered = df[df['client_id'] == most_entries]['qty'].sum()
units_ordered

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [76]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
shrunk_df = df[['unit_price', 'qty', 'subtotal']]
shrunk_df.head(3)

unit_price  qty  subtotal
11110      175.21  100  17521.00
151        125.47  139  17440.33
5052        30.21   61   1842.81

In [77]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['total_weight'] = df['unit_weight'] * df['qty']

df['shipping_price'] = np.where(df['total_weight'] > 50, df['total_weight'] * 7, df['total_weight'] * 10)

weight_price_df = df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']]
weight_price_df.head(3)


unit_price  unit_weight  qty  total_weight  shipping_price
11110      175.21        13.83  100       1383.00         9681.00
151        125.47         0.54  139         75.06          525.42
5052        30.21         2.62   61        159.82         1118.74

In [78]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df.rename(columns = {"subtotal":"line_subtotal"}, inplace = True)
sales_tax = ((df['line_subtotal'] + df['shipping_price']) * .0925).round(2)
df['total_price'] = df['line_subtotal'] + df['shipping_price'] + sales_tax
total_price_df = df[['line_subtotal', 'shipping_price', 'total_price']]
total_price_df.head(3)


line_subtotal  shipping_price  total_price
11110       17521.00         9681.00     29718.18
151         17440.33          525.42     19627.58
5052         1842.81         1118.74      3235.49

In [79]:
df.rename(columns = { "total_price" : "line_price"}, inplace = True)
df


first    last                                       job  \
11110  Samantha  Murray                             Herpetologist   
151    Samantha  Murray                             Herpetologist   
5052   Samantha  Murray                             Herpetologist   
31669  Samantha  Murray                             Herpetologist   
42337  Samantha  Murray                             Herpetologist   
...         ...     ...                                       ...   
38158     David   Ochoa  Geographical information systems officer   
48120     David   Ochoa  Geographical information systems officer   
53449     David   Ochoa  Geographical information systems officer   
38447     David   Ochoa  Geographical information systems officer   
44111     David   Ochoa  Geographical information systems officer   

                    phone                                    email  client_id  \
11110  534.390.2260x56734  murray.samantha.4079@gonzalez-brown.com      99984   
151    534.390.2260x56734  murray.samantha.4079@gonzalez-brown.com      99984   
5052   534.390.2260x56734  murray.samantha.4079@gonzalez-brown.com      99984   
31669  534.390.2260x56734  murray.samantha.4079@gonzalez-brown.com      99984   
42337  534.390.2260x56734  murray.samantha.4079@gonzalez-brown.com      99984   
...                   ...                                      ...        ...   
38158        281.867.1615     david.measure.3427@morales-young.org      10033   
48120        281.867.1615     david.measure.3427@morales-young.org      10033   
53449        281.867.1615     david.measure.3427@morales-young.org      10033   
38447        281.867.1615     david.measure.3427@morales-young.org      10033   
44111        281.867.1615     david.measure.3427@morales-young.org      10033   

       order_id  order_date  order_week  order_year  ...        subcategory  \
11110   9673005  2023-02-27           9        2023  ...  bathroom supplies   
151     6821353  2023-02-17           7        2023  ...             tables   
5052    9673005  2023-02-27           9        2023  ...             stands   
31669   8437912  2023-01-25           4        2023  ...   kitchen supplies   
42337   7810921  2023-01-16           3        2023  ...     paper products   
...         ...         ...         ...         ...  ...                ...   
38158   8159112  2023-01-16           3        2023  ...             stands   
48120   3229787  2023-01-30           5        2023  ...       audio visual   
53449   9244566  2023-03-11          10        2023  ...             stands   
38447   9244566  2023-03-11          10        2023  ...   kitchen supplies   
44111   9244566  2023-03-11          10        2023  ...         accounting   

      unit_price unit_cost  unit_weight   qty  line_number  line_subtotal  \
11110     175.21    162.27        13.83   100            2       17521.00   
151       125.47     86.77         0.54   139            8       17440.33   
5052       30.21     19.49         2.62    61            0        1842.81   
31669       8.67      6.33         2.22    53            0         459.51   
42337     101.40     74.52         8.37  6145            1      623103.00   
...          ...       ...          ...   ...          ...            ...   
38158       1.48      1.22        18.04   146            3         216.08   
48120     136.99    105.43         2.47    60            0        8219.40   
53449     166.92    104.79         1.49   287            7       47906.04   
38447     155.31    146.56         0.04    37            2        5746.47   
44111     310.36    200.41        21.39   287            5       89073.32   

       total_weight  shipping_price  line_price  
11110       1383.00         9681.00    29718.18  
151           75.06          525.42    19627.58  
5052         159.82         1118.74     3235.49  
31669        117.66          823.62     1401.82  
42337      51433.65       360035.55  1074078.87  
...             ...             ..

In [80]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df.head()


first    last            job               phone  \
11110  Samantha  Murray  Herpetologist  534.390.2260x56734   
151    Samantha  Murray  Herpetologist  534.390.2260x56734   
5052   Samantha  Murray  Herpetologist  534.390.2260x56734   
31669  Samantha  Murray  Herpetologist  534.390.2260x56734   
42337  Samantha  Murray  Herpetologist  534.390.2260x56734   

                                         email  client_id  order_id  \
11110  murray.samantha.4079@gonzalez-brown.com      99984   9673005   
151    murray.samantha.4079@gonzalez-brown.com      99984   6821353   
5052   murray.samantha.4079@gonzalez-brown.com      99984   9673005   
31669  murray.samantha.4079@gonzalez-brown.com      99984   8437912   
42337  murray.samantha.4079@gonzalez-brown.com      99984   7810921   

       order_date  order_week  order_year  ... unit_price unit_cost  \
11110  2023-02-27           9        2023  ...     175.21    162.27   
151    2023-02-17           7        2023  ...     125.47     86.77   
5052   2023-02-27           9        2023  ...      30.21     19.49   
31669  2023-01-25           4        2023  ...       8.67      6.33   
42337  2023-01-16           3        2023  ...     101.40     74.52   

      unit_weight   qty  line_number  line_subtotal  total_weight  \
11110       13.83   100            2       17521.00       1383.00   
151          0.54   139            8       17440.33         75.06   
5052         2.62    61            0        1842.81        159.82   
31669        2.22    53            0         459.51        117.66   
42337        8.37  6145            1      623103.00      51433.65   

       shipping_price  line_price  line_cost  
11110         9681.00    29718.18   25908.00  
151            525.42    19627.58   12586.45  
5052          1118.74     3235.49    2307.63  
31669          823.62     1401.82    1159.11  
42337       360035.55  1074078.87  817960.95  

[5 rows x 23 columns]

In [81]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['line_price'] - df['line_cost']
df.head()


first    last            job               phone  \
11110  Samantha  Murray  Herpetologist  534.390.2260x56734   
151    Samantha  Murray  Herpetologist  534.390.2260x56734   
5052   Samantha  Murray  Herpetologist  534.390.2260x56734   
31669  Samantha  Murray  Herpetologist  534.390.2260x56734   
42337  Samantha  Murray  Herpetologist  534.390.2260x56734   

                                         email  client_id  order_id  \
11110  murray.samantha.4079@gonzalez-brown.com      99984   9673005   
151    murray.samantha.4079@gonzalez-brown.com      99984   6821353   
5052   murray.samantha.4079@gonzalez-brown.com      99984   9673005   
31669  murray.samantha.4079@gonzalez-brown.com      99984   8437912   
42337  murray.samantha.4079@gonzalez-brown.com      99984   7810921   

       order_date  order_week  order_year  ... unit_cost unit_weight   qty  \
11110  2023-02-27           9        2023  ...    162.27       13.83   100   
151    2023-02-17           7        2023  ...     86.77        0.54   139   
5052   2023-02-27           9        2023  ...     19.49        2.62    61   
31669  2023-01-25           4        2023  ...      6.33        2.22    53   
42337  2023-01-16           3        2023  ...     74.52        8.37  6145   

       line_number  line_subtotal  total_weight  shipping_price  line_price  \
11110            2       17521.00       1383.00         9681.00    29718.18   
151              8       17440.33         75.06          525.42    19627.58   
5052             0        1842.81        159.82         1118.74     3235.49   
31669            0         459.51        117.66          823.62     1401.82   
42337            1      623103.00      51433.65       360035.55  1074078.87   

       line_cost  line_profit  
11110   25908.00      3810.18  
151     12586.45      7041.13  
5052     2307.63       927.86  
31669    1159.11       242.71  
42337  817960.95    256117.92  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [82]:
# Check your work using the totals above
receipt_data = [{'Order ID' : 2742071,
               'Total Price' : 152811.89
              },
              {'Order ID' : 2173913,
               'Total Price': 162388.71
              },
            {'Order ID' : 6128929,
             'Total Price': 923441.25
            }]
receipt_df = pd.DataFrame(receipt_data)
receipt_df
              



Order ID  Total Price
0   2742071    152811.89
1   2173913    162388.71
2   6128929    923441.25

In [83]:
# Check your work using the totals above
compar_df = df[['order_id', 'line_price']]
compar_df
              



order_id  line_price
11110   9673005    29718.18
151     6821353    19627.58
5052    9673005     3235.49
31669   8437912     1401.82
42337   7810921  1074078.87
...         ...         ...
38158   8159112    20378.36
48120   3229787    10113.05
53449   9244566    55607.65
38447   9244566     6294.19
44111   9244566   144260.07

[54639 rows x 2 columns]

In [84]:
# Check your work using the totals above
compar_df[compar_df['order_id'].isin([2742071, 2173913, 6128929])]
              



order_id  line_price
14517   6128929   180327.76
38542   6128929      173.84
16295   6128929     4133.06
13672   6128929     8121.21
37949   6128929   367345.26
21141   6128929     1713.39
48834   6128929   102182.66
3422    6128929    48763.65
49353   6128929   208535.46
37265   6128929     2144.96
24671   2742071     1976.73
23840   2742071     1273.61
20623   2742071     2069.02
6287    2742071   135889.52
42470   2742071     4973.32
3446    2742071     6629.69
14962   2173913     5076.03
27713   2173913     4297.41
14089   2173913      538.03
627     2173913     9856.86
22004   2173913     9640.08
14938   2173913     7092.74
53155   2173913   125887.56

In [85]:
# Check your work using the totals above
order_2742071 = compar_df[compar_df['order_id'] == 2742071]['line_price'].sum()
order_2173913 = compar_df[compar_df['order_id'] == 2173913]['line_price'].sum()
order_6128929 = compar_df[compar_df['order_id'] == 6128929]['line_price'].sum()



In [86]:
# Check your work using the totals above
print(f'Order 2742071 total: ${order_2742071}')
print(f'Order 2173913 total: ${order_2173913}')
print(f'Order 6128929 total: ${order_6128929}')



Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [87]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
clients_5

client_id
33615    220
66037    211
46820    209
24741    207
38378    207
Name: count, dtype: int64

In [88]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

unique_client = clients_5.unique()
print(unique_client)


[220 211 209 207]


In [89]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

c_5_list

[33615, 66037, 46820, 24741, 38378]

In [90]:
for client_id in c_5_list:
    total_spending = df.loc[df['client_id'] == client_id, 'qty'].sum() * \
    df.loc[df['client_id'] == client_id, 'line_price'].values[0]
    print(f' {client_id} -- ${total_spending:.2f}')

 33615 -- $315989062.90
 66037 -- $425026013.42
 46820 -- $2510744673.36
 24741 -- $1149536236.38
 38378 -- $5527134086.21


In [117]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

summary_data = []
for client_id in c_5_list:
    clients_5_df = df[df['client_id'] == client_id]
    units_purchased_5 = clients_5_df['qty'].sum()
    shipping_price_5 = clients_5_df['shipping_price'].sum()

    
    # revenue_5 = clients_5_df['line_price'].sum() * units_purchased_5
    # cost_5 = (clients_5_df['unit_cost'].sum() * units_purchased_5) + shipping_price_5
    # profit_5 = revenue_5 - cost_5
    revenue_5 = clients_5_df['line_price'].sum()
    cost_5 = clients_5_df['line_cost'].sum()
    profit_5 = clients_5_df['line_profit'].sum()

    summary_data.append({'client_id': client_id,
                                        'total_units' : units_purchased_5,
                                        'total_shipping_price' : shipping_price_5,
                                        'total_revenue' : revenue_5,
                                        'total_cost' : cost_5,
                                        'total_profit' : profit_5
                                       })
summary_df_clients5 = pd.DataFrame(summary_data)
summary_df_clients5

client_id  total_units  total_shipping_price  total_revenue   total_cost  \
0      33615        64313            1828984.89     8377308.52   6175313.91   
1      66037        43018            1395151.85    10259514.79   7004482.98   
2      46820        75768            1601448.84     9743794.36   7007191.64   
3      24741       239862            5126448.37    82268892.02  45688899.71   
4      38378        73667            3429455.40    12906550.87   9634720.98   

   total_profit  
0    2201994.61  
1    3255031.81  
2    2736602.72  
3   36579992.31  
4    3271829.89

In [118]:
# Format the data and rename the columns to names suitable for presentation.
summary_df_clients5.rename(columns={
    'client_id' : 'Client ID',
    'total_units' : 'Units',
    'total_shipping_price' : 'Shipping (millions)',
    'total_revenue' : 'Total Revenue (millions)',
    'total_cost' : 'Total Cost (millions)',
    'total_profit' : 'Total Profit (millions)'
}, inplace = True)
summary_df_clients5
# Define the money columns. 
money_columns = ['Shipping (millions)', 'Total Revenue (millions)', 'Total Cost (millions)', 'Total Profit (millions)']
# Define a function that converts a dollar amount to millions.
def currency_format_millions(amount):
    
    return amount / 1000000

# Apply the currency_format_millions function to only the money columns. 
summary_df_clients5[money_columns] = summary_df_clients5[money_columns].apply(currency_format_millions)
# Rename the columns to reflect the change in the money format. 
summary_df_clients5



Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      33615   64313             1.828985                  8.377309   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
3      24741  239862             5.126448                 82.268892   
4      38378   73667             3.429455                 12.906551   

   Total Cost (millions)  Total Profit (millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3              45.688900                36.579992  
4               9.634721                 3.271830

In [119]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
summary_df_clients5.sort_values(by ='Total Profit (millions)', ascending = False)

Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
3      24741  239862             5.126448                 82.268892   
4      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
3              45.688900                36.579992  
4               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995

In [ ]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.


Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             5.126448                 82.268892   
3      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995